backend-api/src/functions/adapters/kmsAdapter.ts

```diff
@@ -19,13 +28,24 @@ export class KMSAdapter implements IKmsAdapter {
     encryptedData: Uint8Array,
     encryptionKeyId: string,
   ): Promise<Uint8Array> {
-    const decryptCommandOutput: DecryptCommandOutput = await kmsClient.send(
-      new DecryptCommand({
-        KeyId: encryptionKeyId,
-        CiphertextBlob: encryptedData,
-        EncryptionAlgorithm: "RSAES_OAEP_SHA_256",
-      }),
-    );
+    let decryptCommandOutput: DecryptCommandOutput;
+    try {
+      decryptCommandOutput = await kmsClient.send(
+        new DecryptCommand({
+          KeyId: encryptionKeyId,
+          CiphertextBlob: encryptedData,
+          EncryptionAlgorithm: "RSAES_OAEP_SHA_256",
+        }),
+      );
+    } catch (error) {
+      if (
+        error instanceof InvalidCiphertextException ||
+        error instanceof IncorrectKeyException
```

Comment:
KMS seems to throw three 500 server errors and around eight 400 errors. Is it worth doing the reverse here and checking for server errors first?

Reply:
@JacksonJ2W can I check re those errors you're referring to (I've not looked at the docs)... Do those KMS 500 and 400 codes directly map to what we would return in the response? Could we could get a bad request from a client and then give KMS something bad which could cause a KMS 500 error but we'd still want to return 400? Or vice versa.

Reply:
That would be the easiest and nicest solution, but the issue is that not all client exceptions thrown by KMS mean that our client (caller) did something wrong. For example, the `InvalidKeyUsageException` is a client exception but it would only be thrown if we (i.e. async backend) make a mistake in the decrypt request we send to KMS, so it would be wrong to return our client a 400 when we made the mistake ourselves and not them.

Reply:
This is a confusing one because in this situation we are KMS' client but we also have a client of our own that we need to respond to...

Reply:
I would rather return a 4XX in the active session response when we defo know the client did something wrong (as is)

Reply:
But we are now, no? 🤔

Reply:
@anahickmann yes sorry that's what I meant by: (as is)

Reply:
Ah ok, sorry 🤦‍♀️

Reply:
No, bad wording from me 😄